File: 2020-week12/c++/2020-week12_5.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

int getBudgets(int left, int right, vector<int> budgets, int M){
    int tmp = 0;
    
    while(left <= right){      
        long long sum = 0;       
        int mid = (left+right)/2;
        
        for(int bg:budgets){            
            bg > mid ? sum += mid : sum += bg;
        }
        
        if(sum > M){
            right = mid-1;
        }
        else {
            left = mid+1;
            tmp = max(tmp, mid);
        }
    }
    return tmp;
}
int solution(vector<int> budgets, int M) {
    
    sort(budgets.begin(), budgets.end(), greater<int>());
    
    int left = 0;
    int right = budgets[0];
    return getBudgets(left, right, budgets, M);
}

File: 2020-week12/js/2020-week12_5.js
function solution(budgets, M) {
  let left = 0;
  let right = M;

  let max = 0;
  let sum = 0;
  for (let i = 0; i < budgets.length; i++) {
    sum += budgets[i];
    max = Math.max(max, budgets[i]);
  }

  if (sum <= M) return max;

  while (left <= right) {
    let mid = Math.floor((left + right) / 2);

    let sum = 0;
    for (let i = 0; i < budgets.length; i++) {
      if (budgets[i] <= mid) sum += budgets[i];
      else sum += mid;
    }

    if (sum <= M) {
      left = mid + 1;
    } else {
      right = mid - 1;
    }
  }

  return right;
}